### DISCUSSION TITLE
Streamlines not generating from 3d data

### DISCUSSION THREAD

**Discussion by dustykeyboard1** [pyvista/pyvista#5269] [Q&A]
I am working on visualizing a 3D vector field in PyVista and encountering difficulties generating streamlines. My data consists of two numpy arrays: one for point locations (new_B, shape (244225, 3)) and another for vector components (new_B_values, shape (244225, 3)).

I'm using the following code to create glyphs:

`point_cloud = pv.PolyData(new_B)`
`point_cloud["vectors"] = new_B_values`
`glyphs = point_cloud.glyph(orient="vectors", tolerance=0.03)`

However, when I try to visualize these glyphs or generate streamlines, I'm running into issues. For instance, when plotting the streamlines:

`streamlines, src = point_cloud.streamlines('vectors', return_source=True, terminal_speed=0.0, n_points=200, source_radius=0.1)`
`streamlines.tube(radius=0.0015).plot(cpos=[(1.2, 1.2, 1.2), (0.0, 0.0, 0.0), (0.0, 0.0, 1.0)])`

I receive an error: ValueError: Empty meshes cannot be plotted. Input mesh has zero points.

I suspect this might be related to the vector field magnitude or the streamline generation parameters. I've tried scaling up the vector magnitudes and adjusting n_points and source_radius, but the issue persists.

Any guidance on how to correctly generate and visualize these streamlines would be greatly appreciated!

**Comment by banesullivan** [comment]
It would help us greatly if you could produce a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help you debug this. Otherwise, this seems like something may be going away with the parameters you have set in the streamline filter for this specific dataset.

Are you also unable to plot the glyphs on their own?

**Reply by dustykeyboard1** [reply]
I am able to plot the glyphs no problem. Its hard to create a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) because of the data. I have added a clearer explanation below.

**Comment by MatthewFlamm** [comment]
Your `point_cloud` is just that, a point cloud of vertex cells.  This is a 0D cell type that does not fill space.  `streamlines` requires 2D or 3D cells.  You need to construct a mesh that has either 2D or 3D cells.  See #5234 for similar question.

**Reply by dustykeyboard1** [reply]
I have tried creating an unstructured grid with the code below:

`points = new_B`

`cells = np.hstack([[1, i] for i in range(len(points))]).astype(np.int_)`

`cell_type = np.full(len(points), pv.CellType.VERTEX)`

`grid = pv.UnstructuredGrid(cells, cell_type, points)`

`grid["vectors"] = new_B_values`

then creating streamlines with this:
`
streamlines, src = unstructured_grid.streamlines(
    return_source=True, terminal_speed=0.0, n_points=200, source_radius=0.1
)
`

then I try plotting with this:

`p = pv.Plotter()`
`p.add_mesh(streamlines.tube(radius=0.0015), lighting=False)`
`p.add_mesh(src)`
`p.show()`

However, I get this error:

{
	"name": "ValueError",
	"message": "Empty meshes cannot be plotted. Input mesh has zero points.",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb Cell 16 line 3
     <a href='vscode-notebook-cell:/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb#Y145sZmlsZQ%3D%3D?line=33'>34</a> # Visualize the streamlines
     <a href='vscode-notebook-cell:/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb#Y145sZmlsZQ%3D%3D?line=34'>35</a> p = pv.Plotter()
---> <a href='vscode-notebook-cell:/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb#Y145sZmlsZQ%3D%3D?line=35'>36</a> p.add_mesh(streamlines.tube(radius=0.0015), lighting=False)
     <a href='vscode-notebook-cell:/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb#Y145sZmlsZQ%3D%3D?line=36'>37</a> p.add_mesh(src)
     <a href='vscode-notebook-cell:/Volumes/USB31FD/MADNT/MAD_research/NewVisualizations/creating3d_black_hole.ipynb#Y145sZmlsZQ%3D%3D?line=37'>38</a> p.show()

File ~/Desktop/Coding/School/Physics/SeniorYear/BlackHoleEnvironment/lib/python3.10/site-packages/pyvista/plotting/plotter.py:3458, in BasePlotter.add_mesh(self, mesh, color, style, scalars, clim, show_edges, edge_color, point_size, line_width, opacity, flip_scalars, lighting, n_colors, interpolate_before_map, cmap, label, reset_camera, scalar_bar_args, show_scalar_bar, multi_colors, name, texture, render_points_as_spheres, render_lines_as_tubes, smooth_shading, split_sharp_edges, ambient, diffuse, specular, specular_power, nan_color, nan_opacity, culling, rgb, categories, silhouette, use_transparency, below_color, above_color, annotations, pickable, preference, log_scale, pbr, metallic, roughness, render, component, emissive, copy_mesh, backface_params, show_vertices, **kwargs)
   3454         raise ValueError('RGB array must be n_points/n_cells by 3/4 in shape.')
   3456 if algo is None and not mesh.n_points:
   3457     # Algorithms may initialize with an empty mesh
-> 3458     raise ValueError('Empty meshes cannot be plotted. Input mesh has zero points.')
   3460 # set main values
   3461 self.mesh = mesh

ValueError: Empty meshes cannot be plotted. Input mesh has zero points."
}

For reference:

new_B is a numpy array with shape `(523777, 3)`. This contains XYZ coordinates for each point. new_B_values is a numpy array with shape  `(523777, 3)`. This contains the vector components for each respective point in new_B. 

Here is a few examples from my data:
<img width="438" alt="Screenshot 2023-12-04 at 12 37 52 PM" src="https://github.com/pyvista/pyvista/assets/78929863/0bc1e3d6-92e7-44d1-ab9c-046bdbfa5899">

**Reply by MatthewFlamm** [reply]
You still have a mesh with only vertex cells.  If your points are truly a point cloud and there is no mesh structure, I would look into the `interpolate` filter which can interpolate onto a mesh with structure.  Or use an algorithm to make a 3D tetra mesh from your point cloud.

**Reply by dustykeyboard1** [reply]
If my points are unstructured, would unstructured grid work with streamlines? From what I can gather, here would be the steps:
1 - create a point cloud with the locations of the points
2 - create an unstructured grid with the point cloud
3 - add the vector values to the grid
4 - interpolate the grid onto a mesh
5 - create streamlines from the mesh. 

Is this correct?

**Reply by MatthewFlamm** [reply]
An `UnstructuredGrid` in the pyvista and vtk sense is generic.  It is a dataset type that is very flexible.  It can represent point clouds (vertex cells), unstructured meshes with cellular topology in 2D and 3D (e.g. triangular mesh in 2D), and structured meshes (although without optimization of memory/processing speed).  For streamlines, it is not enough to have an `UnstructuredGrid`.  You need to have a mesh that has cells that are 2D or 3D.

Are you obtaining these points from a simulation that has a mesh?  If so, this would be the place to start, i.e. obtaining the mesh structure from the simulation.

If you truly have a point cloud, I would modify your list like this:

1 - create a point cloud with the locations of the points
3 - add the vector values to the point cloud
4 - interpolate the grid onto a mesh
5 - create streamlines from the mesh.

See https://docs.pyvista.org/version/stable/examples/01-filter/interpolate.html

**Reply by dustykeyboard1** [reply]
The data is from the HARM black hole simulation. It does not have a mesh.

**Reply by dustykeyboard1** [reply]
So I have generated a single streamline from this data:

`
# Create a point cloud
point_cloud = pv.PolyData(new_B)
point_cloud["vectors"] = new_B_values

grid_x_min, grid_x_max = new_B[:, 0].min(), new_B[:, 0].max()
grid_y_min, grid_y_max = new_B[:, 1].min(), new_B[:, 1].max()
grid_z_min, grid_z_max = new_B[:, 2].min(), new_B[:, 2].max()
grid_resolution = [100, 100, 100]

# Create a grid
grid = pv.UniformGrid()
grid.dimensions = grid_resolution
grid.origin = (grid_x_min, grid_y_min, grid_z_min)
grid.spacing = (
    (grid_x_max - grid_x_min) / (grid_resolution[0] - 1),
    (grid_y_max - grid_y_min) / (grid_resolution[1] - 1),
    (grid_z_max - grid_z_min) / (grid_resolution[2] - 1),
)

# Interpolate point cloud data onto the grid
grid = grid.interpolate(point_cloud, radius=3, sharpness=2)

# Generate streamlines
streamlines, src = grid.streamlines(
    "vectors", return_source=True, terminal_speed=0.0, n_points=200, source_radius=0.1
)

# Visualize the streamlines
p = pv.Plotter()
p.add_mesh(streamlines.tube(radius=0.0015), lighting=False)
p.add_mesh(src)
p.show()
`

However, it is only generating one streamline, with a changing value in vector. I was under the impression this would generate multiple streamlines from the vectors that are points in the relatively same direction.

**Reply by MatthewFlamm** [reply]
My quick reading of HARM code is that it has a connectivity structure that defines the transport fluxes, but these equations are outside my expertise area.

Check that all your starting points are inside your mesh.  Also, when interpolating, you need to be careful that you tune the parameters carefully to avoid artifacts.

**Reply by MatthewFlamm** [reply]
Also try a line source for the streamlines, it may be more obvious what is going on.  I'm not a fan of the default sphere source.

**Reply by dustykeyboard1** [reply]
I got it to work! Genertating plenty of streamlines working. Not sure if its correct but, at least it is working. I have one final question - do you know of a way that I can plot the stream lines in napari. I was able to do when interpolating the denisty values. With code such as this:


```
point_cloud = pv.PolyData(new_rho)
point_cloud["density"] = flattened_rho_data
centroid = np.mean(new_rho, axis=0)
grid = pv.create_grid(point_cloud)
grid = grid.interpolate(point_cloud, radius=4, sharpness=5)

```

and 
```
volume_data = grid["density"].reshape(grid.dimensions)
viewer.add_image(volume_data, colormap="hsv", translate=centroid/2, opacity=.5)
```


I was curious if you knew of a similar way to do this for the streamlines? Thanks for your help!

**Reply by MatthewFlamm** [reply]
I'm not familiar with napari.  It sounds like a question for napari.  But this is all doable to render in pyvista directly.

**Reply by dustykeyboard1** [reply]
Yes I know. I unfortunately have to use napari. Thank you for your help. I will close this.
